### PR TITLE
fix(api-sdk): updated cq directory handling code

### PR DESCRIPTION
**Pull request by RamilGaripov**
refs. metriport/metriport-internal#1350

### Dependencies

- Upstream: https://github.com/metriport/metriport/pull/1565
- Upstream: https://github.com/metriport/metriport-internal/pull/1491

### Description

- Updated the CQDirectoryModel to include 2 new fields: `gateway` (to indicate whether an org is a gateway or not, and `managing_organization`
- Added migration files to update the table
- Improved OID extraction regex
- Added code to post-process CQ directory rebuilding to find all gateways in the directory
- Added CQ XCPD prep logic that prevents requests to CW

### Testing

- Local
  - [x] Local testing of both staging and prod directories

### Release Plan

- :warning: This contains a DB migration
- [ ] Upstream dependencies are met/released
- [ ] Release NPM packages
- [ ] Merge this

### CODE REVIEW

File: packages/shared/src/common/normalize-oid.ts [hunk 1]
```diff
@@ -1,10 +1,25 @@
-const OID_REGEX = /^(urn:oid:)?([0-9]+(\.[0-9]+)*)$/;
+const OID_REGEX = /(?:[^.\d]*)((([0-9]*)(?:\.*))*)*(?:[^.\d]*)/;
```
Comment (leite08) [inline packages/shared/src/common/normalize-oid.ts:1]: Why the change here, to be more lenient w/ the first chars (i.e., allow for typos)?

It seems the regex can be significantly simplified (no need for all those matching groups IMO, and shouldn't we check for start/end of the string?

Wouldn't this suffice (maybe adding start/end of string)?

```
[^.\d]*:([\d.]+)
```

File: packages/api/src/external/carequality/command/cq-directory/create-cq-directory-entry.ts [hunk 2]
```diff
@@ -4,6 +4,7 @@ import { CQDirectoryEntryModel } from "../../models/cq-directory";
 import { cqDirectoryEntryTemp } from "./shared";
 
 const keys = createKeys();
+const NUMBER_OF_KEYS = keys.split(",").length;
```
[the remark this answers is not in the document]
Comment (RamilGaripov) [reply packages/api/src/external/carequality/command/cq-directory/create-cq-directory-entry.ts:7]: ah, it shouldnt be

File: packages/api/src/external/carequality/command/cq-directory/cq-gateways.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+      gateway: true,
+    },
+    {
+      where: {
+        name: {
+          [Op.like]: `%${gateway}%`,
+        },
+        urlXCPD: {
+          [Op.ne]: "",
+        },
```
[the remark this answers is not in the document]
Comment (RamilGaripov) [reply packages/api/src/external/carequality/command/cq-directory/cq-gateways.ts:21]: Yea, I guess I had to use an empty string because of the other code you pointed out where I was assigning them [empty strings](https://github.com/metriport/metriport/pull/1575#discussion_r1486333939). Will fix this. 

The logic for not updating them is that gateways will be organizations like athena, epic, etc - large structures that we can send PD requests to. If an organization doesn't have `urlXCPD`, it is not a gateway

File: packages/api/src/external/carequality/command/cq-directory/cq-gateways.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+export async function updateCQGateway(gateway: string): Promise<void> {
+  await CQDirectoryEntryModel.update(
+    {
+      gateway: true,
+    },
+    {
+      where: {
+        name: {
+          [Op.like]: `%${gateway}%`,
```
[the remark this answers is not in the document]
Comment (RamilGaripov) [reply packages/api/src/external/carequality/command/cq-directory/cq-gateways.ts:17]: Yea, I think just the one at the end of `name` is sufficient in this case

File: packages/api/src/external/carequality/command/cq-directory/parse-cq-directory-entry.ts [hunk 5]
```diff
@@ -86,24 +85,28 @@ function getUrls(contained: Contained): XCUrls | undefined {
...
-    urls.urlDR = endpointMap[XCA_DR_STRING];
+  if (isValidUrl(urlDR)) {
+    urls.urlDR = urlDR;
   }
 
   return urls;
 }
 
+function isValidUrl(url: string | undefined): boolean {
+  return url ? url.startsWith("http") : false;
```
[the remark this answers is not in the document]
Comment (RamilGaripov) [reply packages/api/src/external/carequality/command/cq-directory/parse-cq-directory-entry.ts:107]: you mean to try and create a URL with it like:
```
function isValidUrl(url: string | undefined): boolean {
  if (!url) return false;

  try {
    new URL(url);
    return true;
  } catch (e) {
    return false;
  }
}
```

File: packages/carequality-sdk/src/models/organization.ts [hunk 6]
```diff
@@ -11,6 +11,11 @@ export const organizationIdentifierSchema = z.object({
   value: objectValue,
 });
 
+export const managinOrganizationSchema = z
+  .object({ reference: z.object({ value: z.string().nullish() }) })
```
[the remark this answers is not in the document]
Comment (RamilGaripov) [reply packages/carequality-sdk/src/models/organization.ts:15]: hmmm.. tbh I missed that part. But I didn't run into any issues while executing the code as is

File: packages/infra/config/env-config.ts [hunk 7]
```diff
@@ -45,6 +45,7 @@ export type EnvConfig = {
   carequality?: {
     secretNames?: {
       CQ_API_KEY?: string;
+      CW_XCPD_URL_IN_CQ?: string;
```
[the remark this answers is not in the document]
Comment (RamilGaripov) [reply packages/infra/config/env-config.ts:48]: I'm not sure CW's XCPD url is available in public domain, and I didn't want to be the one to expose it

File: packages/shared/src/common/normalize-oid.ts [hunk 8]
```diff
@@ -1,10 +1,25 @@
...
+    }
   }
 
   throw new Error("OID is not valid");
 }
+
+function startsWithRootCode(oid: string): boolean {
+  if (oid.includes(".")) {
+    const root = oid.split(".")[0];
+    return root ? parseInt(root) <= 2 : false;
```
[the remark this answers is not in the document]
Comment (RamilGaripov) [reply packages/shared/src/common/normalize-oid.ts:20]: This scenario isn't possible in the context of the caller function, which will parse out the `-` sign. But good point, if we want to use this function elsewhere. Fixed.

File: packages/shared/src/common/normalize-oid.ts (hunk 1 above)
Comment (RamilGaripov) [reply packages/shared/src/common/normalize-oid.ts:1]: I had to get a little creative with this string because it seems that CQ directory has no standards for what the OID should look like. There's some wild edge cases, typos, etc inside the value for OID, and I didn't wanna miss out on any of them that contain legit OIDs.

Comment (leite08) [reply packages/shared/src/common/normalize-oid.ts:1]: Can we add those do a unit test that validate they pass this regex? This way we can update the regex in the future if needed, and know it will still work with those OIDs.

File: packages/api/src/external/carequality/command/cq-directory/search-cq-directory.ts [hunk 9]
```diff
@@ -2,18 +2,21 @@ import { Patient } from "@metriport/core/domain/patient";
 import { Coordinates } from "@metriport/core/external/aws/location";
 import convert from "convert-units";
 import { Sequelize } from "sequelize";
+import { Config } from "../../../../shared/config";
 import { CQDirectoryEntryModel } from "../../models/cq-directory";
 
 export const DEFAULT_RADIUS_IN_MILES = 50;
+const cqIgnoreList: string[] = constructGatewayIgnoreList();
```
Comment (jonahkaye) [inline packages/api/src/external/carequality/command/cq-directory/search-cq-directory.ts:9]: hyper nit: I would call this exclude not ignore. Ignore would be like requests coming in and we ignore them , but doesnt make as much sense to say ignore when talking about us choosing orgs to send requests to

File: packages/api/src/external/carequality/organization-conversion.ts [hunk 10]
```diff
@@ -3,11 +3,14 @@ import { XCPDGateways } from "@metriport/ihe-gateway-sdk";
...
-      id: uuidv7(),
-      oid: org.id,
-      url: org.urlXCPD,
-    };
+  return cqOrgs.flatMap(org => {
+    if (org.urlXCPD) {
+      return {
+        url: org.urlXCPD,
+        oid: org.id,
+        id: uuidv7(),
```
Comment (jonahkaye) [inline packages/api/src/external/carequality/organization-conversion.ts:11]: if this uuid is random and is generated per request, it should be generated in the logic of `prepareForPatientDiscovery` or `createPatientDiscoveryRequest` and not here in a more generic conversion function

File: packages/api/src/sequelize/migrations/2024-02-05_00_alter-cq-directory-add-gateway.ts [hunk 11]
```diff
@@ -0,0 +1,58 @@
+import { DataTypes } from "sequelize";
+import type { Migration } from "..";
+
+const tableName = "cq_directory_entry";
+const gatewayColumn = "gateway";
+const managingOrgColumn = "managing_organization";
+const activeColumn = "active";
+const columnToChange = "url_xcpd";
+
+// Use 'Promise.all' when changes are independent of each other
```
Comment (jonahkaye) [inline packages/api/src/sequelize/migrations/2024-02-05_00_alter-cq-directory-add-gateway.ts:10]: leftover?

File: packages/shared/src/common/normalize-oid.ts (hunk 1 above)
Comment (jonahkaye) [reply packages/shared/src/common/normalize-oid.ts:1]: ✕ should throw an error for non-numeric characters in OID parts
  ✕ should throw an error for OIDs with spaces or other delimiters
  
  I ran a few unit tests with a few more boundary conditions. Will push commit to branch but two cases we arent handling right now (unless we are and this is expected behavior)

Comment (RamilGaripov) [reply packages/shared/src/common/normalize-oid.ts:1]: Thanks, ya, that's good. I was focusing to account for all of the existing cases in the CQ directory where we would've previously missed an OID. But further improvements are always welcome

Comment (jonahkaye) [reply packages/shared/src/common/normalize-oid.ts:1]: yah that makes sense then to not handle those cases if the source of OIDs is like a directory we can see. My thinking was just that because it was in a shared package, the logic was supposed to be more general, but your approach makes sense and no need to over engineer

**Review by jonahkaye — APPROVED** [review]
Re: regex and unit testing LGTM

File: packages/carequality-sdk/src/client/carequality.ts [hunk 12]
```diff
@@ -80,12 +87,29 @@ export class CarequalityManagementAPIImpl implements CarequalityManagementAPI {
...
       baseURL: baseUrl,
       httpsAgent: this.httpsAgent,
     });
+
+    axiosRetry(this.api, {
+      retries: options?.retries ?? DEFAULT_MAX_RETRIES,
+      retryDelay: retryCount => {
+        const exponentialDelay = Math.pow(2, retryCount);
+        const jitter = Math.random();
+        const delayWithJitter = (exponentialDelay + jitter) * BASE_DELAY.asMilliseconds();
```
Comment (leite08) [inline packages/carequality-sdk/src/client/carequality.ts:105]: Not a huge deal, but assuming `retryCount` starts at `1`, `BASE_DELAY` is not the initial delay - which one would assume that's what it means without reading the code.

This is because `Math.pow(2, retryCount)` will start at `2`. We could instead do `Math.pow(2, Math.max(0, retryCount - 1))`

File: packages/carequality-sdk/src/client/carequality.ts [hunk 13]
```diff
@@ -80,12 +87,29 @@ export class CarequalityManagementAPIImpl implements CarequalityManagementAPI {
...
+        return Math.min(delayWithJitter, this.maxBackoff);
+      },
+      retryCondition: error => {
+        return error.response?.status !== 200;
```
Comment (leite08) [inline packages/carequality-sdk/src/client/carequality.ts:109]: Don't we only want to retry on error? 2xx is success, 3xx is redirection - [reference](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes).

Also, do we want to retry on 4xx? It would be useful to have a link to the CQ spec indicating this behavior, maybe on the PR description or on Notion and a link from here to it?

File: packages/carequality-sdk/src/client/carequality.ts [hunk 14]
```diff
@@ -80,12 +87,29 @@ export class CarequalityManagementAPIImpl implements CarequalityManagementAPI {
...
+      ? dayjs.duration(options.maxBackoff, "seconds").asMilliseconds()
+      : DEFAULT_MAXIMUM_BACKOFF.asMilliseconds();
 
     this.api = axios.create({
       timeout: options?.timeout ?? DEFAULT_AXIOS_TIMEOUT.asMilliseconds(),
       baseURL: baseUrl,
       httpsAgent: this.httpsAgent,
     });
+
+    axiosRetry(this.api, {
```
Comment (leite08) [inline packages/carequality-sdk/src/client/carequality.ts:100]: We tested the retry/backoff behavior, right?

File: packages/shared/src/common/__tests__/normalize-oid.test.ts
Comment (leite08) [inline packages/shared/src/common/__tests__/normalize-oid.test.ts:1]: love this! 😍 

NABD in this PR, but ideally each test case would test one scenario.

File: packages/api/src/sequelize/migrations/2024-02-05_00_alter-cq-directory-add-gateway.ts
Comment (leite08) [inline packages/api/src/sequelize/migrations/2024-02-05_00_alter-cq-directory-add-gateway.ts:1]: NABD, but if you send a new commit could consider updating the date on this migration so it matches the day its being merged.

File: packages/api/src/external/carequality/command/cq-directory/search-cq-directory.ts [hunk 15]
```diff
@@ -2,18 +2,21 @@ import { Patient } from "@metriport/core/domain/patient";
...
 export const DEFAULT_RADIUS_IN_MILES = 50;
+const cqExcludeList: string[] = constructGatewayExcludeList();
 
 export type CQOrgBasicDetails = {
   name: string | undefined;
   id: string;
   lon: number | undefined;
   lat: number | undefined;
-  urlXCPD: string;
+  urlXCPD: string | undefined | null;
```
Comment (leite08) [inline packages/api/src/external/carequality/command/cq-directory/search-cq-directory.ts:16]: Can we not use `null` here? Could convert from `undefined` to `null` if needed - [reference](https://www.notion.so/metriport/Development-Best-Practices-bb3574d3c83d46dc9914a9e5c5542ca7?pvs=4#51c298a38cba4897a9fa639218e69833).

File: packages/api/src/external/carequality/command/cq-directory/cq-gateways.ts [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+  );
+}
+
+export async function getCQGateways(): Promise<CQDirectoryEntryModel[]> {
+  return CQDirectoryEntryModel.findAll({
+    where: {
+      gateway: true,
+      urlXCPD: {
+        [Op.ne]: "",
+      },
```
Comment (leite08) [inline packages/api/src/external/carequality/command/cq-directory/cq-gateways.ts:43]: Can't we just use `gateway: true` as a way to indicate which ones are GW?

Also, the logic here is not the same as the `updateCQGateway()` fn, which uses `null` to identify GWs. This is exactly why we should try to keep logic in one place.

**Review by leite08 — CHANGES_REQUESTED** [review]
Also please check unanswered comments from prior review

File: packages/api/src/external/carequality/command/cq-directory/cq-gateways.ts [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+import { capture } from "@metriport/core/util/notifications";
+
+export async function updateCQGateways(gateways: string[]): Promise<void> {
+  console.log(`Found ${gateways.length} gateways in the CQ directory. Updating...`);
+  await Promise.allSettled(
+    gateways.map(gateway =>
+      updateCQGateway(gateway).catch(error => {
+        const msg = `Failed to update gateway ${gateway} in the CQ directory. Skipping...`;
+        console.log(`${msg}. Cause: ${error}`);
+        capture.error(msg, { extra: { context: `cq.directory.updateCQGateways`, error } });
```
Comment (leite08) [inline packages/api/src/external/carequality/command/cq-directory/cq-gateways.ts:12]: Please add `gateway` to `extra`

File: packages/api/src/external/carequality/command/cq-directory/parse-cq-directory-entry.ts [hunk 18]
```diff
@@ -86,24 +87,35 @@ function getUrls(contained: Contained): XCUrls | undefined {
     }
   });
 
+  const urls: XCUrls = {};
   const urlXCPD = endpointMap[XCPD_STRING];
+  const urlDQ = endpointMap[XCA_DQ_STRING];
+  const urlDR = endpointMap[XCA_DR_STRING];
 
-  if (!urlXCPD) return;
```
Comment (leite08) [inline packages/api/src/external/carequality/command/cq-directory/parse-cq-directory-entry.ts:91]: Now that we don't return `undefined` anymore, let's update the function's return type?

File: packages/api/src/external/carequality/cq-directory.ts [hunk 19]
```diff
@@ -4,14 +4,17 @@ import { Organization } from "@metriport/carequality-sdk/models/organization";
 export type CQDirectoryEntryData = {
   id: string; // Organization's OID
   name?: string;
-  urlXCPD: string;
+  urlXCPD?: string | null;
```
Comment (leite08) [inline packages/api/src/external/carequality/cq-directory.ts:7]: ditto: avoid `null`

File: packages/api/src/external/carequality/models/cq-directory.ts [hunk 20]
```diff
@@ -10,15 +10,18 @@ export class CQDirectoryEntryModel
   static NAME = "cq_directory_entry";
   declare id: string; // Organization's OID
   declare name?: string;
-  declare urlXCPD: string;
+  declare urlXCPD?: string | null;
```
Comment (leite08) [inline packages/api/src/external/carequality/models/cq-directory.ts:13]: I'd think `null` is needed here since this is the model and this column is nullable. But other columns have optional/undefined, so maybe we can also avoid `null` here?

**Review by leite08 — APPROVED** [review]
Approved considering no changes on how we deal w/ CQ gateways.

**Review by Goncharo — APPROVED** [review]
:shipit:
